[AArch64] Update IssueWidth for Neoverse V1, N1, N3

simonwallis2 opened the pull request:
Recently the IssueWidth in the scheduling model was reduced for Neoverse-V2 and N2. This patch does the same for Neoverse-V1, N1 and N3.

On Neoverse-V1, various values of IssueWidth (15, 8, 7, 6, 5) were tried with runs of various workloads.
The highest overall geomean score was achieved with an issue width of 8. No significant regressions were noted.

On Neoverse-N1, various values of IssueWidth (8, 6, 5, 4, 3) were tried with runs of various workloads.
The highest overall geomean score was achieved with an issue width of 3. No significant regressions were noted.

On Neoverse-N3, it makes sense to do exactly the same as was done for N2. It is proposed to use an issue width of 5.

Related V2 PR: https://github.com/llvm/llvm-project/pull/142565
Related N2 PR: https://github.com/llvm/llvm-project/pull/145717

Commit:
[AArch64] Update IssueWidth for Neoverse V1, N1, N3

Recently the IssueWidth in the scheduling model was reduced for Neoverse-V2 and N2.
This patch does the same for Neoverse-V1, N1 and N3.

On Neoverse-V1, various values of IssueWidth (15, 8, 7, 6, 5)
were tried with runs of various workloads.
The highest overall geomean score was achieved with an issue width of 8.
No significant regressions were noted.

On Neoverse-N1, various values of IssueWidth (8, 6, 5, 4, 3)
were tried with runs of various workloads.
The highest overall geomean score was achieved with an issue width of 3.
No significant regressions were noted.

On Neoverse-N3, it makes sense to do exactly the same as was done for N2.
It is proposed to use an issue width of 5.

Related V2 PR: https://github.com/llvm/llvm-project/pull/142565
Related N2 PR: https://github.com/llvm/llvm-project/pull/145717

Change-Id: I7d17359e7a4ae7a527e321a208bdf96893fe50f9

davemgreen commented:
This sounds sensible. What was the difference between 3 and 4 for neoverse-n1? A value of 4 would help match the SWOG (and #136374), and if there isn't much in it would make a better value I believe.

simonwallis2 commented:
> This sounds sensible. What was the difference between 3 and 4 for neoverse-n1? A value of 4 would help match the SWOG (and #136374), and if there isn't much in it would make a better value I believe.

For neoverse-n1, with a value of 4 I saw some small amount of noise but no overall improvement over the original value of 8.

c-rhodes commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseN1.td:
```diff
@@ -15,7 +15,7 @@
 
 def NeoverseN1Model : SchedMachineModel {
-  let IssueWidth            =   8; // Maximum micro-ops dispatch rate.
+  let IssueWidth            =   3; // Maximum micro-ops dispatch rate.
```
can we update these comments to match the v2 scheduler following this discussion: https://github.com/llvm/llvm-project/pull/142565#discussion_r2126798020

simonwallis2 replied:
Sure. Done for AArch64SchedNeoverseN1.td and AArch64SchedNeoverseV1.td.

Commit:
[AArch64] Update IssueWidth for Neoverse V1, N1, N3

Recently the IssueWidth in the scheduling model was reduced for Neoverse-V2 and N2.
This patch does the same for Neoverse-V1, N1 and N3.

On Neoverse-V1, various values of IssueWidth (15, 8, 7, 6, 5)
were tried with runs of various workloads.
The highest overall geomean score was achieved with an issue width of 8.
No significant regressions were noted.

On Neoverse-N1, various values of IssueWidth (8, 6, 5, 4, 3)
were tried with runs of various workloads.
The highest overall geomean score was achieved with an issue width of 3.
No significant regressions were noted.

On Neoverse-N3, it makes sense to do exactly the same as was done for N2.
It is proposed to use an issue width of 5.

Related V2 PR: https://github.com/llvm/llvm-project/pull/142565
Related N2 PR: https://github.com/llvm/llvm-project/pull/145717

Change-Id: I7d1e71e0470407bfd49b535a1068ff52235e9792

davemgreen reviewed (approved):
I would go with a IssueWidth of 4 for NeoverseN1, as I believe that matches reality better, but either way this LGTM.

c-rhodes reviewed (approved):
Also LGTM, cheers

simonwallis2 commented:
In llvm/tools/llvm-mca, there are 10 instances of 
format("%.2f", …)

9 of these instances explicitly round up, using
floor((X * 100) + 0.5) / 100)

This RT output is the 10th instance. 
llvm-mca/InstructionInfoView.cpp should be modified to match the other 9, and the test cases updated to match the new rounded-up value.
